src/useObservable.ts

```diff
@@ -33,24 +38,40 @@ export function useObservable<T>(observable: Observable<T>): T | undefined
 export function useObservable<T>(observable: Observable<T>, initialValue: T): T
 export function useObservable<T>(observable: Observable<T>, initialValue: () => T): T
 export function useObservable<T>(observable: Observable<T>, initialValue?: T | (() => T)) {
-  const [getSnapshot, subscribe] = useMemo(() => {
+  const [getSnapshot, subscribe] = useMemo<
+    [() => T, Parameters<typeof useSyncExternalStore>[0]]
+  >(() => {
     const record = getOrCreateStore(observable, getValue(initialValue))!
     return [
       function getSnapshot() {
+        // @TODO: perf opt opportunity: we could do `record.subscription.unsubscribe()` here to clear up some memory, as this subscription is only needed to provide a sync initialValue.
         return record.currentValue
       },
-      function subscribe(callback: (value: T) => void) {
-        const sub = record.observable.subscribe(next => callback(next))
+      function subscribe(callback: () => void) {
+        // @TODO: perf opt opportunity: we could do `record.subscription.unsubscribe()` here as we only need 1 subscription active to keep the observer alive
+        const sub = record.observable.subscribe(() => callback())
         return () => {
           sub.unsubscribe()
         }
       },
     ]
   }, [observable])
 
-  useIsomorphicEffect(() => {
```

Comment:
As this side-effect is only needed to clear up memory after unmounting it's not necessary to use `useIsomorphicEffect`/`useLayoutEffect`. The sync subscription ensures that, if possible, `record.currentValue` will have a value on first render and we don't have to pay the perf cost of `useLayoutEffect` (with its `render + block browser paint + layout effect + render + allow browser paint` flow).

Reply:
Great insight, thank you!